Ngrx component setup (#2046)
ngrx/platform: projects/ngrx.io/content/guide/component/index.md

Also, other well-known libraries, [angular material](https://material.angular.io/) provide a reactive way of usage.

This creates even more interest and for so-called `reactive primitives` for the Angular framework, like the `async` and other template syntax, decorators and services.
 
This library provides a ser of primitives making it more convenient to work in a reactive architecture.
  
## Installation 

Detailed installation instructions can be found on the [Installation](guide/component/install) page.

## Included Primitives

<!-- toc -->

- [Push Pipe](#push-pipe)
- [Let Structural Directive](#let-structural-directive)
- [Observable Life Cycle Hooks](#observable-life-cycle-hooks)
  * [selectChanges RxJS Operator](#selectchanges-rxjs-operator)
- [Observable Input Bindings](#observable-input-bindings)
- [Observable Output Bindings](#observable-output-bindings)
- [Local State Management](#local-state-management)
  * [selectSlices RxJS Operator](#selectslices-rxjs-operator)

<!-- tocstop -->

### Push Pipe

> An angular pipe similar to the `async` pipe but triggers `detectChanges` instead of `markForCheck`.

Reviewer: ```suggestion
An Angular pipe similar to the `async` pipe but triggers `detectChanges` instead of `markForCheck`.
```

This is required to run zone-less. We render on every pushed message.
(currently, there is an [isssue](https://github.com/angular/angular/issues/31438) with the `ChangeDetectorRef` in ivy so we have to wait for the fix.

The pipe should work as template binding `{{thing$ | push}}` 

> as well as input binding `[color]="thing$ | push"` and trigger the changes of the host component.

Reviewer: ```suggestion
as well as an input binding `[color]="thing$ | push"` and triggers the changes of the host component.
```

```html
<div *ngIf="(thing$ | push) as thing">
  color: {{thing.color}}
  shape: {{thing.shape}}
<div>

<app-color [color]="(thing$ | push).color">
</app-color>
```

**Included Features:**

> - subscription handling overview  life cycle

Reviewer: ```suggestion
- subscription handling overview life cycle
```

> - a unified way of handling null and undefined with streams

Reviewer: ```suggestion
- a unified way of handling `null` and `undefined` values with streams
```

- optional flag to turn off scheduling over `AnimationFrameScheduler` (on by default)
- change detection is done manually which allows it to work zone-less too

### Let Structural Directive

The `*rgnxLet` directive serves a convenient way of binding multiple observables in the same view context.
It also helps with some extra features under the hood.

> The current way of handling subscriptions in the view looks like that:

Reviewer: ```suggestion
The current way of handling subscriptions in the view looks like this:
```

```html
<ng-container *ngIf="{
              color: observable1$ | async,
              shape: observable2$ | async,
              name:  observable3$ | async
            } as c">
  {{c.color}}
  <app-color [color]="c.color" [shape]="c.shape" [name]="c.name">
  </app-color>  
</ng-container>
```

The `*ngrxLet` directive take over several things and makes it more convenient and save to work with streams in the template
`*ngrxLet="o$ as o;"` 

```html
<!-- observables = { color: observable1$, shape: observable2$, name:  observable3$ } -->

<ng-container *ngrxLet="observables$ as c">
  <app-color [color]="c.color" [shape]="c.shape" [name]="c.name">
  </app-color>
</ng-container>

<ng-container *ngrxLet="observables; let c">
  <app-color [color]="c.color" [shape]="c.shape" [name]="c.name">
  </app-color>
</ng-container>

<ng-container *ngrxLet="observables; color as c; shape as s; name as n">
  <app-color [color]="c" [shape]="s" [name]="n">
  </app-color>
</ng-container>
```

**Included Features:**
- binding is always present. (`*ngIf="{}"` normally effects it)
- it takes away the multiple usages of the `async` pipe 

> - proper handling of null and undefined values

Reviewer: ```suggestion
- proper handling of `null` and `undefined` values
```

- removes state slices if bound observable completes or errors
- an option to disable scheduling over `AnimationFrameScheduler` (on by default)
- control change detection and therefore can run zone-less

### Observable Life Cycle Hooks

A property decorator which turns a lifecycle method into an observable and assigns it to the related property.

The decorator should work as a proxy for all life cycle hooks ` @hook$('onInit') onInit$;` 
as well as forward passed values i.e. `changes` in from the `OnChanges` hook.

```typescript
  @hook$('onInit') onInit$;
  @hook$('onDestroy') onDestroy$;

  this.onInit$
    .pipe(
      switchMapTo(interval(1000)),
      map(_ => Date.now()),
      takeUntil(this.onDestroy$)
    )
    .subscribe();
```

**Included Features**

> - it handles late subscribers.

Reviewer: ```suggestion
- it handles late subscribers
```

- exposes only observables
- respects single shot vs ongoing life cycles
- subscription handling over the component lifetime
- return the latest value when resubscribing

#### selectChanges RxJS Operator

> An operators `selectChanges` to select one or many specific slices from `SimpleChange`. 

Reviewer: ```suggestion
An operator `selectChanges` to select one or many specific slices from `SimpleChange`. 
```

This operator can be used in combination with `onChanges$`.

It also provides a very early option to control the forwarded values.

**Example of selectSlice operator**
```typescript
export class MyComponent {
  @hook$('onChanges') 
  onChanges$: Observable<SimpleChanges>;

  @Input() state;
  state$ = this.onChanges$.pipe(getChange('state'));
}
```

Following things are done under the hood:
- pull out `currentValue` from `SimpleChanges` object

> - optional it could have a parma for a custom comparison function

Reviewer: ```suggestion
- optional it could have a param for a custom comparison function
```

### Observable Input Bindings

> A property decorator which turns component or directive input binding into an observable and assigned it to the related property.

Reviewer: ```suggestion
A property decorator which turns component or directive input binding into an observable and assigns it to the related property.
```

```typescript
@Component({
  selector: 'app-child',
  template: `<p>input: {{input$ | async}}</p>`,
})
export class ChildComponent  {
  @Input$()
  input$;
}
```

Following things are done under the hood:
- It caches to consider late subscribers (life cycle hook related) 
- It is multi-casted to avoid multiple subscriptions
- It works with WebComponents and AngularComponents

### Observable Output Bindings

TBD from Dominic Elm

### Local State Management

> This extension is maybe the most interesting one. While we can 

Reviewer: I think it's missing something here?
While we can ...

A tiny logic that combines:
- values over input bindings
- The component class internal state
- state rendered to view 
- state from services.

> A way to connect events from the view and component as observable.

Reviewer: ```suggestion
A way to connect events from the view and component as observables.
```

```typescript
constructor(private lS: LocalState<MyState>) {
  this.lS
    .connectSlice('num', interval(1000));
  this.lS
    .connectSlice('isNew', this.isNew$);
  this.lS
    .connectSlice('buttons', this.buttons$);
}
```

#### selectSlices RxJS Operator

A flexible way to query one or many state slices.

> It considers also a late subscriber. 

Reviewer: ```suggestion
It also considers late subscribers. 
```

> An operators `selectSlice` to select a specific slice from the managed state. 

Reviewer: ```suggestion
An operator `selectSlices` to select a specific slice from the managed state. 
```

This operator can be used to get slices from `this.lS$`.

```typescript
buttons$ = this.lS.state$
  .pipe(

>     selectChange(['state', 'substate'])

Reviewer: `selectSlices` ?

  );
```

Following things are done under the hood:
- it handles late subscribers with `shareReplay(1)` 

> - it forwards only distinct values

Reviewer: ```suggestion
- it only forwards distinct values
```

# @ngrx/component

Reactive Component/Decorator library for Angular applications.

> Included are decorators and directives that support a fully reactive in Angular Components.  

Reviewer 2: Do you mean a reactive `state`?

```suggestion
Included are decorators and directives that support a fully reactive state in Angular Components.  
```

## Introduction

> Parts of Angular like the `ReactiveFromsModule`, `RouterModule`, `HttpClientModule`, etc. are already reactive.

Reviewer 2: ```suggestion
Parts of Angular like the `ReactiveFormsModule`, `RouterModule`, `HttpClientModule`, etc. are already reactive.
```

For those who prefer imperative code, it's a little effort to restrict it to a single subscription.

On the other hand for those who prefer reactive code, it's not that easy. 
A lot of conveniences are missing, and beside the `async` pipe there is pretty much nothing to take away the manual mapping to observables.  
Furthermore, an increasing number of packages start to be fully observable based. A very popular and widely used example is [NgRx](https://ngrx.io/). It enables us to maintain global push-based state management based on observables. 

> Also, other well-known libraries, [angular material](https://material.angular.io/) provide a reactive way of usage.

Reviewer 2: ```suggestion
Also, other well-known libraries, [Angular material](https://material.angular.io/) provide a reactive way of usage.
```

The pipe should work as template binding `{{thing$ | push}}` 
as well as an input binding `[color]="thing$ | push"` and triggers the changes of the host component.

```html
<div *ngIf="(thing$ | push) as thing">
  color: {{thing.color}}
  shape: {{thing.shape}}
<div>

<app-color [color]="(thing$ | push).color">
</app-color>
```

**Included Features:**
- subscription handling overview life cycle
- a unified way of handling `null` and `undefined` values with streams
- optional flag to turn off scheduling over `AnimationFrameScheduler` (on by default)
- change detection is done manually which allows it to work zone-less too

### Let Structural Directive

The `*ngrxLet` directive serves a convenient way of binding multiple observables in the same view context.
It also helps with some extra features under the hood.

The current way of handling subscriptions in the view looks like this:

```html
<ng-container *ngIf="{
              color: observable1$ | async,
              shape: observable2$ | async,
              name:  observable3$ | async
            } as c">
  {{c.color}}
  <app-color [color]="c.color" [shape]="c.shape" [name]="c.name">
  </app-color>  
</ng-container>
```

> The `*ngrxLet` directive take over several things and makes it more convenient and save to work with streams in the template

Reviewer 2: ```suggestion
The `*ngrxLet` directive takes over several things and makes it more convenient and safe to work with streams in the template
```

> `*ngrxLet="o$ as o;"` 

Reviewer 2: ```suggestion
`*ngrxLet="o$ as o"` 
```

The decorator should work as a proxy for all life cycle hooks ` @hook$('onInit') onInit$;` 
as well as forward passed values i.e. `changes` in from the `OnChanges` hook.

```typescript
  @hook$('onInit') onInit$;
  @hook$('onDestroy') onDestroy$;

  this.onInit$
    .pipe(
      switchMapTo(interval(1000)),
      map(_ => Date.now()),
      takeUntil(this.onDestroy$)
    )
    .subscribe();
```

**Included Features**
- it handles late subscribers
- exposes only observables
- respects single shot vs ongoing life cycles
- subscription handling over the component lifetime
- return the latest value when resubscribing

#### selectChanges RxJS Operator

An operator `selectChanges` to select one or many specific slices from `SimpleChange`. 
This operator can be used in combination with `onChanges$`.

It also provides a very early option to control the forwarded values.

**Example of selectSlice operator**
```typescript
export class MyComponent {
  @hook$('onChanges') 
  onChanges$: Observable<SimpleChanges>;

  @Input() state;
  state$ = this.onChanges$.pipe(getChange('state'));
}
```

Following things are done under the hood:
- pull out `currentValue` from `SimpleChanges` object

> - optional it could have a param for a custom comparison function

Reviewer 2: ```suggestion
- optional: it could have a param for a custom comparison function
```

Parts of Angular like the `ReactiveFormsModule`, `RouterModule`, `HttpClientModule`, etc. are already reactive.
For those who prefer imperative code, it's a little effort to restrict it to a single subscription.

On the other hand for those who prefer reactive code, it's not that easy. 
A lot of conveniences are missing, and beside the `async` pipe there is pretty much nothing to take away the manual mapping to observables.  
Furthermore, an increasing number of packages start to be fully observable based. A very popular and widely used example is [NgRx](https://ngrx.io/). It enables us to maintain global push-based state management based on observables. 
Also, other well-known libraries, [Angular material](https://material.angular.io/) provide a reactive way of usage.

This creates even more interest and for so-called `reactive primitives` for the Angular framework, like the `async` and other template syntax, decorators and services.
 
This library provides a set of primitives making it more convenient to work in a reactive architecture.
  
## Installation 

Detailed installation instructions can be found on the [Installation](guide/component/install) page.

## Included

> - [Push Pipe](#guide/component/push)

Reviewer: ```suggestion
- [Push Pipe](guide/component/push)
```

In order to build the docs

> - [Let Structural Directive](#guide/component/let)

Reviewer: ```suggestion
- [Let Structural Directive](guide/component/let)
```

Same
